Microsoft.Toolkit.Uwp.Input.GazeInteraction/GazePointer.cs

```diff
@@ -0,0 +1,918 @@
+                _currentlyFixatedElement != null)
+            {
+                _currentlyFixatedElement.Invoke();
+            }
+        }
+
+        /// <summary>
+        /// Run device calibration.
+        /// </summary>
+        /// <returns>Task that returns true, if calibration completes successfully; otherwise, false.</returns>
+        public Task<bool> RequestCalibrationAsync()
+        {
+            return _devices.Count == 1 ?
+                _devices[0].RequestCalibrationAsync().AsTask() :
+                Task.FromResult(false);
+        }
+
+        /// <summary>
+        /// Event handler for Gaze related events, of with a <see cref="GazeEventArgs"/> object.
+        /// </summary>
+        public event EventHandler<GazeEventArgs> GazeEvent
+        {
+            add
+            {
+                _gazeEventCount++;
+                _gazeEvent += value;
+            }
+
+            remove
+            {
+                _gazeEventCount--;
+                _gazeEvent -= value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the UIElement representing the cursor.
+        /// </summary>
+        public UIElement CursorElement
+        {
+            get { return _gazeCursor.PopupChild; }
+            set { _gazeCursor.PopupChild = value; }
+        }
+
+        private EventHandler<GazeEventArgs> _gazeEvent;
+
+        private int _gazeEventCount = 0;
+
+        internal Brush EnterBrush { get; set; } = null;
+
+        internal Brush ProgressBrush { get; set; } = new SolidColorBrush(Colors.Green);
+
+        internal Brush CompleteBrush { get; set; } = new SolidColorBrush(Colors.Red);
+
+        internal double DwellStrokeThickness { get; set; } = 2;
+
+        internal Interaction Interaction { get; set; } = Interaction.Disabled;
+
+        internal GazeTargetItem NonInvokeGazeTargetItem { get; }
+
+        internal GazeFeedbackPopupFactory GazeFeedbackPopupFactory { get; } = new GazeFeedbackPopupFactory();
+
+        internal void Reset()
+        {
+            _activeHitTargetTimes.Clear();
+            _gazeHistory.Clear();
+
+            _maxHistoryTime = DEFAULT_MAX_HISTORY_DURATION;
+        }
+
+        internal void SetElementStateDelay(UIElement element, PointerState relevantState, TimeSpan stateDelay)
+        {
+            var property = GetProperty(relevantState);
+            element.SetValue(property, stateDelay);
+
+            // fix up _maxHistoryTime in case the new param exceeds the history length we are currently tracking
+            var dwellTime = GetElementStateDelay(element, PointerState.Dwell);
+            var repeatTime = GetElementStateDelay(element, PointerState.DwellRepeat);
+            _maxHistoryTime = new TimeSpan(2 * Math.Max(dwellTime.Ticks, repeatTime.Ticks));
+        }
+
+        /// <summary>
+        /// Find the parent to inherit properties from.
+        /// </summary>
+        private static UIElement GetInheritenceParent(UIElement child)
```

Comment:
I see this visual tree walking pattern a lot, is there ways to use our TreeHelpers or improve them and centralize some of this code???

Then we could try and focus on optimizing the TreeHelpers in general...?

Reply:
And add a dependency on the main nuget just for this? I'm not convinced yet, but maybe... Lets track this somewhere else.